course3/course_en.md

```diff
@@ -0,0 +1,724 @@
+  - Its implementation is not our main focus, and it can be repleced with `@vec.Vector[Option[Int64]]`.
+
+---
+
+# Top-Down Implementation
+
+- Before each computation, we first check if our desired result has been cached.
+  - If it does, we can simply use the result.
+  - If it doesn't, we calculate the result and store it in the data structure.
+
+```moonbit expr
+fn fib1(num: Int) -> Int64 {
+  fn aux(num: Int, map: @map.Map[Int, Int64]) -> (Int64, @map.Map[Int, Int64]) {
+    match get(map, num) {
+      Some(result) => (result, map)
+      None => {
+        let (result_1, map_1) = aux(num - 1, map)
+        let (result_2, map_2) = aux(num - 2, map_1)
+        (result_1 + result_2, put(map_2, num, result_1 + result_2))
+      }
+    }
+  }
+  let map = put(put(make(), 1, 1L), 2, 1L)
+  aux(num, map).0
+}
+```
+
+---
+
+# Mutable Variables
+
+Note that `map` is being updated and passed as an argument repeatedly.
+To simplify the code, MoonBit supports mutable variables.
+
+```moonbit expr
+fn fib1_mut(num: Int) -> Int64 {
+  // Declare a mutable variable with let mut
+  let mut map = put(put(make(), 1, 1L), 2, 1L)
+  fn aux(num: Int) -> Int64 {
+    match get(map, num) {
+      Some(result) => result
+      None => {
+        let result_1 = aux(num - 1)
+        let result_2 = aux(num - 2)
+        // Update the binding with <variable> = <expression>
+        map = put(map, num, result_1 + result_2) 
+        result_1 + result_2
+      }
+    }
+  }
+  aux(num)
+}
+```
+
+---
+
+# Bottom-Up Implementation
+
+Starting from the first one, calculate and store the subsequent items sequentially.
+
+```moonbit expr
+fn fib2(num: Int) -> Int64 {
+  fn aux(n: Int, map: @map.Map[Int, Int64]) -> Int64 {
+    let result = get_or_else(get(map, n - 1), 1L) + 
+      get_or_else(get(map, n - 2), 1L)
+    if n == num { result } 
+    else { aux(n + 1, put(map, n, result)) }
+  }
+  let map = put(put(make(), 0, 0L), 1, 1L)
+  aux(1, map)
+}
+```
+
+---
+
+# Bottom-Up Implementation
+
+- Since we only need the results of the previous two subproblems when computing the current one, we can pass the results directly through the parameters, without the use of additional data structures.
+
+```moonbit expr
+fn fib2(num : Int) -> Int64 {
+  fn aux(n: Int, acc1: Int64, acc2: Int64) -> Int64 {
+    match n {
+      0 => acc1
+      _ => aux(n - 1, acc2, acc1 + acc2)
+    }
+  }
+  aux(num, 0L, 1L)
+}
+```
+
+![](../pics/fib_aux.drawio.png)
+
+---
+
+# Summary
+
+- In this module we learned
+   - Basic data type: functions and their operations
+   - Data structure: lists and pattern matching on lists
+   - Algorithm: recursions and dynamic programming
+- Extended reading:
+   - _Software Foundations, Volume 1: Logical Foundations_
+     Basics, Induction & Lists
+   - _Programming Language Foundations in Agda_
+     Naturals, Induction & Relations
+   - _Introduction to Algorithms_
+     Chapter 15 (3e) / Chapter 14 (4e): Dynamic Programming
```

Comment:
The 4th edition of CLRS was recently published in 2022, and many people currently hold the 3rd edition, so listing the chapter numbers of different editions together with the chapter titles can avoid misunderstandings.